src/Future.js

```diff
@@ -82,6 +84,79 @@ Future.prototype.chain = function(f) {  // Sorella's:
   }.bind(this));
 };
 
+var chainRecFork = function(t, rej, res) {
+  var isSync = false;
+  t.fork(function(v) {
+    var r = rej(v, !isSync);
+    isSync = true;
+    return r;
+  }, function(v) {
+    var r = res(v, !isSync);
+    isSync = true;
+    return r;
+  });
+  if (isSync) {
+    return { isSync: true };
+  } else {
+    isSync = true;
+    return { isSync: false };
+  }
+};
+
+// chainRec
+Future.chainRec = Future.prototype.chainRec = function(f, i) {
```

Comment:
There are a lot of moving parts in here which is proving to be a bit difficult for me to follow. I think perhaps some comments in the code could help out a lot here.

Alternatively, the implementation of the MonadRec instance for Aff translates quite nicely.
```js
// Heavily influenced by the Aff MonadRec instance
// https://github.com/slamdata/purescript-aff
function tailRecM(f, a) {
  var UNRESOLVED = 0,
      SYNC       = 1,
      ASYNC      = 2;

  return Future(function(reject, resolve) {
    function go(acc) {
      var result, status;
      while (true) {
        status = UNRESOLVED;
        f(Left, Right, acc).fork(reject, function _resolve(v) {
          if (status === UNRESOLVED) {
            // If this callback has been called while the status is still
            // UNRESOLVED then this has been a synchronous call. If this
            // were asynchronous, then the status would have been updated
            // below before this callback was invoked.
            status = SYNC;
            // Store the result for further synchronous processing.
            result = v;
          } else {
            // Otherwise it is safe to call recursively due to async callback.
            Either.either(go, resolve, v);
          }
        });

        // status will be SYNC if the above callback function has already been invoked.
        if (status === SYNC) {
          if (Either.isLeft(result)) {
            // update the value for the next loop and continue
            acc = result.value;
            continue;
          } else {
            // otherwise we're done
            resolve(result.value);
          }
        } else {
          // If the status hasn't been updated to SYNC yet, it must be ASYNC.
          status = ASYNC;
        }
        // async futures will exit the while loop here
        return;
      }
    }
    return go(a);
  });
}
```